src/cmap/establish/handshake/mod.rs

```diff
@@ -184,18 +194,82 @@ impl Handshaker {
     }
 
     /// Handshakes a connection.
-    pub(super) async fn handshake(&self, conn: &mut Connection) -> Result<()> {
-        let response = conn.send_command(self.command.clone(), None).await?;
-        let command_response = response.body()?;
+    pub(super) async fn handshake(
+        &self,
+        conn: &mut Connection,
+        credential: Option<&Credential>,
+    ) -> Result<Option<FirstRound>> {
+        let mut command = self.command.clone();
+
+        // Update the command document with the speculative authentication info.
+        let client_first = set_client_first(&mut command.body, credential)?;
+
+        let start_time = PreciseTime::now();
+        let response = conn.send_command(command, None).await?;
+        let end_time = PreciseTime::now();
+
+        let mut command_response: IsMasterCommandResponse = response.body()?;
+
+        // Record the client's message and the server's response from speculative authentication if
+        // the server did send a response.
+        let first_round = client_first.and_then(|client_first| {
+            command_response
+                .speculative_authenticate
+                .take()
+                .map(|server_first| {
+                    let command_response = CommandResponse::with_document_and_address(
+                        conn.address().clone(),
+                        server_first,
+                    );
+
+                    client_first.into_first_round(command_response)
+                })
+        });
 
-        // TODO RUST-192: Calculate round trip time.
```

Comment:
This TODO should have been done a while ago, so I just did it here. We were properly doing this in the monitoring checks, so it's not a huge deal that we didn't have it here, but it's worth having given that this is the round trip that will be exposed to users who use a predicate SelectionCriteria.